text/1000-migrations.rst

```diff
@@ -76,6 +80,71 @@ whether there are untracked changes in the local schema that require a
 new migration.
 
 
+Migration History
+=================
+
+The migration history is defined as a linear sequence of migrations (although
+this might change in the future to support merging migration histories).
+Each migration has a unique name that is a hash derived from the token
+stream of the migration script and the migration parent name.  To
+make migration names valid identifiers the hash is prefixed with letter
+``'m'``.
+
+The names of the migrations are computed by the ``edgedb migrate`` command
+starting from the latest committed migration (if any), using the following
+algorithm:
+
+1. Get the latest migration id from the database.
+2. Scans all files in the migrations dir and finds a file that references
+   that parent.
+3. If there's no such file --- terminate: there's no in-progress migration.
+4. if there's one such file and it has the largest sequential file name --
+   then there's only one in-progress migration file, so it is committed.
+5. If there's a file with that parent and other files after it it means
+   that we have a few in-progress (not yet committed) migration files.
+   The parent names in them are fluid -- generated and present in the file,
+   but aren't yet in the database.
+   Scenarios:
+   a) A user has two in-progress migration files: 0004.edgeql and 0005.edgeql.
+      If the user modifies 0004.edgeql and runs edgedb migration commit, we
+      compute the name for 0004 and commit it; we then update the parent name
+      in 0005 and commit it too.
+   b) A user has two in-progress migration files: 0004.edgeql and 0005.edgeql.
+      If the user modifies 0005.edgeql and runs edgedb migration commit,
+      we first commit the 0004, validate that 0005 has the correct parent
+      name, and if needed we fix it. We also compute it's new name, and then
+      commit 00005.
+
+If the tool detects that any of the already-committed migrations have been
+altered, it should issue an error and abort.  This RFC makes no special
+provision as to the recovery from this situation.  The user will either
+have to revert the migration file alteration, or reset the migration history
+with ``RESET SCHEMA``.
+
+
+Migration File Format
+=====================
+
+The file format and layout used by the suite of CLI commands described in
+this RFC is defined as follows.
+
+Each migration is stored in a separate file.  Files are named using a
+simple monotonic decimal counter ('00001.edgeql', '00002.edgeql', etc).
+The contents of the file is the body of the migration and must be valid
+EdgeQL.  Migration name, the name of the parent migration, the
+migration description message and other potential metadata are recorded
+in a comment block at the top of the file:
+
+    # migration: <migration-name>
```

Comment:
Would be cool if we could stick 'autogenerated' somewhere to highlight that `migration` and `parent` fields aren't supposed to be edited.

What about updates to the migration message? Would changes to it be detected and synced to the DB?

Reply:
I think any modification to the already-committed migration should at least be an action that is confirmed by the user, but other than that I see no harm in being able to sync non-critical migration data to the db.

Reply:
>  but other than that I see no harm in being able to sync non-critical migration data to the db.

I wasn't suggesting it's harmful, I was suggesting that we don't know how to easily detect such changes -- the hash does not include the message.

Reply:
> I wasn't suggesting it's harmful, I was suggesting that we don't know how to easily detect such changes -- the hash does not include the message, or does it?

If it does, we need another hash to detect that script is the same and it's safe to update metadata (git by the way has both: hash of contents of the commit (tree) and commit hash with metadata, and this allows certain scenarios easier)

Reply:
I don't think we should include message and other stuff in the hash, only the parent and the script, so an extra hash doesn't seem to be necessary.

Reply:
> so an extra hash doesn't seem to be necessary.

But it is necessary, otherwise we can't detect changes to the migration message on the disk and sync them with the DB.  A metadata hash that @tailhook suggests makes sense to me.

Reply:
> But it is necessary, otherwise we can't detect changes to the migration message

Why?  Read the files and run an introspection query and compare.  Am I missing something?

Reply:
Yeah, alright, I guess it doesn't really matter if we introspect and compare messages vs introspect and compare hashes.

The command line tooling can have its own cache with hashes of all migration files. That would simplify (and greatly speedup) detection of changed files / metadata.